packages/blockchain/src/blockchain.ts

```diff
@@ -369,89 +371,177 @@ export class Blockchain implements BlockchainInterface {
+        //   if (childHeaderHash !== undefined) {
+        //     const childHeader = await this.dbManager
+        //       .getHeaderSafe(childHeaderHash, blockNumber + BIGINT_1)
+        //       .catch((_e) => undefined)
+        //     isHeadChainStillCanonical =
+        //       childHeader !== undefined && equalsBytes(childHeader.parentHash, blockHash)
+        //   } else {
+        //     isHeadChainStillCanonical = false
+        //   }
+        // } else {
+        //   isHeadChainStillCanonical = true
+        // }
 
-        // set total difficulty in the current context scope
-        if (this._headHeaderHash) {
-          currentTd.header = await this.getTotalDifficulty(this._headHeaderHash)
-        }
-        if (this._headBlockHash) {
-          currentTd.block = await this.getTotalDifficulty(this._headBlockHash)
+        let dbOps: DBOp[] = []
+        dbOps = dbOps.concat(DBSetBlockOrHeader(item))
+        dbOps.push(DBSetHashToNumber(blockHash, blockNumber))
+
+        let parentTd = await this.dbManager
+          .getTotalDifficultySafe(header.parentHash, BigInt(blockNumber) - BIGINT_1)
+          .catch((_e) => undefined)
+        const isOptimistic = parentTd === undefined
+        parentTd = parentTd ?? opts?.parentTd
+        const isComplete = parentTd !== undefined
+
+        if (!isOptimistic) {
+          if (this._validateBlocks && !isGenesis && item instanceof Block) {
+            // this calls into `getBlock`, which is why we cannot lock yet
+            await this.validateBlock(block)
+          }
         }
 
-        // calculate the total difficulty of the new block
-        const parentTd = await this.getParentTD(header)
-        if (!block.isGenesis()) {
-          td += parentTd
-        }
+        // 1. if canonical is explicit false then just dump the block
+        // 2. if canonical is explicit true then apply that even for the pow/poa blocks
+        //    if they are optimistic, i.e. can't apply the normal rule
+        // 3. if canonical is not defined, then apply normal rules
+        if (opts?.canonical === false || (opts?.canonical === undefined && !isComplete)) {
+          if (parentTd !== undefined) {
+            const td = header.difficulty + parentTd
+            dbOps = dbOps.concat(DBSetTD(td, blockNumber, blockHash))
+          }
+          await this.dbManager.batch(dbOps)
+        } else {
+          let updatesHead = block.isGenesis()
+          let updatesHeadBlock
+          if (parentTd === undefined) {
+            // if the block is pow and optimistic, and has not been explicity marked canonical, then
+            // throw error since pow blocks can't be optimistically added without expicit instruction about
+            // their canonicality
+            if (
+              !block.isGenesis() &&
+              block.common.consensusType() !== ConsensusType.ProofOfStake &&
+              opts?.canonical === undefined
```

Comment:
if parentTD is undefined, and this is not an explicity specified canonical PoS block, throw error because we can't determine the canonicality of this block

ideally such a scenario will not happen becuase we will not end up in else part of condition of L424, i will do a cleanup and see if we can come to a simplified logic, it could be redundant